runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java

```
@@ -0,0 +1,406 @@
+      return fn.new ProcessContext() {
+        // Commit at least once every 10k output records.  This keeps the watermark advancing
+        // smoothly, and ensures that not too much work will have to be reprocessed in the event of
+        // a crash.
+        // TODO: Also commit at least once every N seconds (runner-specific parameter).
+        private static final int MAX_OUTPUTS_PER_BUNDLE = 10000;
+
+        private int numOutputs = 0;
+
+        public InputT element() {
+          return element.getValue();
+        }
+
+        public Instant timestamp() {
+          return element.getTimestamp();
+        }
+
+        public PaneInfo pane() {
+          return element.getPane();
+        }
+
+        public void output(OutputT output) {
+          baseContext
+              .windowingInternals()
+              .outputWindowedValue(
+                  output, element.getTimestamp(), element.getWindows(), element.getPane());
+          noteOutput();
+        }
+
+        public void outputWithTimestamp(OutputT output, Instant timestamp) {
+          baseContext
+              .windowingInternals()
+              .outputWindowedValue(output, timestamp, element.getWindows(), element.getPane());
+          noteOutput();
+        }
+
+        private void noteOutput() {
+          if (++numOutputs > MAX_OUTPUTS_PER_BUNDLE) {
+            // Request a checkpoint. The fn *may* produce more output, but hopefully not too much.
+            residualRestrictionHolder[0] = tracker.checkpoint();
+          }
+        }
+
+        public <T> T sideInput(PCollectionView<T> view) {
+          return baseContext.sideInput(view);
+        }
+
+        public PipelineOptions getPipelineOptions() {
+          return baseContext.getPipelineOptions();
+        }
+
+        public <T> void sideOutput(TupleTag<T> tag, T output) {
+          // TODO: I'm not sure this is correct, but there's no "internals.sideOutputWindowedValue".
```

Maybe just `throw UnsupportedOperationException("Side outputs not yet supported by SplittableDoFn")`? Which also raises the question, what if the `DoFn` produces more side-outputs than main-outputs? Shouldn't this increase the output counter too?